barman/cloud_providers/aws_s3.py

```diff
@@ -783,14 +805,57 @@ def take_snapshot_backup(self, backup_info, instance_identifier, volumes):
             WaiterConfig=self._get_waiter_config(),
         )
 
+        # Apply lock on snapshots if lock mode is specified
+        if self.lock_mode:
+            self._lock_snapshots(
+                snapshots,
+                self.lock_mode,
+                self.lock_duration,
+                self.lock_cool_off_period,
+                self.lock_expiration_date
+            )
+
         backup_info.snapshots_info = AwsSnapshotsInfo(
             snapshots=snapshots,
             region=self.region,
             # All snapshots will have the same OwnerId so we get it from the last
             # snapshot response.
             account_id=snapshot_resp["OwnerId"],
+
         )
 
+    def _lock_snapshots(self, snapshots, lock_mode, lock_duration, lock_cool_off_period, lock_expiration_date):
+        lock_snapshot_default_args = {
+            "LockMode": lock_mode
+        }
+
+        if lock_duration:
+            lock_snapshot_default_args["LockDuration"] = lock_duration
+
+        if lock_cool_off_period:
+            lock_snapshot_default_args["CoolOffPeriod"] = lock_cool_off_period
+
+        if lock_expiration_date:
+            lock_snapshot_default_args["ExpirationDate"] = lock_expiration_date
+
+        for snapshot in snapshots:
+            lock_snapshot_args = lock_snapshot_default_args.copy()
+            lock_snapshot_args["SnapshotId"] = snapshot.identifier
+
+            resp = self.ec2_client.lock_snapshot(**lock_snapshot_args)
+
+            logging.info(
+                "Snapshot %s locked in state '%s' (lock duration: %s days, cool-off period: %s hours, "
+                "cool-off period expires on: %s, lock expires on: %s, lock duration time: %s)",
+                snapshot.identifier,
+                resp["LockState"],
+                resp["LockDuration"],
+                resp["CoolOffPeriod"],
+                resp["CoolOffPeriodExpiresOn"],
+                resp["LockExpiresOn"],
+                resp["LockDurationStartTime"]
+            )
```

Comment:
In accordance with what @andremagui  said, I think this is the biggest problem here. It seems the response from AWS is not always the same depending on the request parameters you pass, so you should ideally not rely on their response for logging.

E.g. if i run `barman-cloud-backup` with `--aws-snapshot-lock-mode governance --aws-snapshot-lock-duration 1` i get:

`2024-10-23 14:14:14,636 [3122] ERROR: Backup failed uploading data ('CoolOffPeriod')`

Because `CoolOffPeriod` is not present in the response:

`
{'SnapshotId': 'snap-0a9e1d413ef5a9254', 'LockState': 'governance', 'LockDuration': 1, 'LockCreatedOn': datetime.datetime(2024, 10, 23, 14, 14, 14, 594000, tzinfo=tzlocal()), 'LockExpiresOn': datetime.datetime(2024, 10, 24, 14, 14, 14, 594000, tzinfo=tzlocal()), 'LockDurationStartTime': datetime.datetime(2024, 10, 23, 14, 14, 14, 594000, tzinfo=tzlocal()), 'ResponseMetadata': {'RequestId': '91c97b91-9dde-4d38-a652-0607ff9574f8', 'HTTPStatusCode': 200, 'HTTPHeaders': {'x-amzn-requestid': '91c97b91-9dde-4d38-a652-0607ff9574f8', 'cache-control': 'no-cache, no-store', 'strict-transport-security': 'max-age=31536000; includeSubDomains', 'vary': 'accept-encoding', 'content-type': 'text/xml;charset=UTF-8', 'transfer-encoding': 'chunked', 'date': 'Wed, 23 Oct 2024 14:14:14 GMT', 'server': 'AmazonEC2'}, 'RetryAttempts': 0}}
`

The snapshot itself was created and locked but the process errored out and the backup is not completed successfully in Barman.

Reply:
If we want to use a formatted message, we should at least use `resp.get` to return  default values when the keys are not present.